torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -428,105 +484,124 @@ void ProcessGroupAgent::enqueueSend(SendWork work) {
-              // by the sender who has determined the future has timed out.
-              return;
-            }
-            // Use futureInfo before destructing it.
-            fm = futureInfo->second.future_;
-            auto endTime = futureInfo->second.endTime_;
-            futures_.erase(id);
-            // look up the corresponding future by its time out and request ID,
-            // and remove it from the timeouts map
-            auto& futuresAtTime = futureTimeouts_[endTime];
-            auto it = futuresAtTime.find(id);
-            TORCH_INTERNAL_ASSERT(
-                it != futuresAtTime.end(),
-                "Error: could not find future in futureTimeouts map, race condition.");
-            futuresAtTime.erase(it);
-            if (futuresAtTime.empty()) {
-              // remove the key from futureTimeouts_
-              futureTimeouts_.erase(endTime);
+                  getWorkerInfo(fromId),
+                  createExceptionResponse(err->what(), requestId));
             }
+          });
+    }
+  } else if (message.isResponse()) {
+    auto id = message.id();
+    std::shared_ptr<FutureMessage> fm = nullptr;
+    {
+      std::lock_guard<std::mutex> lock{futureMutex_};
+      const auto& futureInfo = futures_.find(id);
+      if (futureInfo == futures_.end()) {
+        // Received a completion for a timed out future, drop the recv.
+        // RecvCounts will not be incremented here, it will be
+        // incremented by the sender who has determined the future has
+        // timed out.
+        return 0;
+      }
+      // Use futureInfo before destructing it.
+      fm = futureInfo->second.future_;
+      auto endTime = futureInfo->second.endTime_;
+      futures_.erase(id);
+      // look up the corresponding future by its time out and request
+      // ID, and remove it from the timeouts map
+      auto& futuresAtTime = futureTimeouts_[endTime];
+      auto it = futuresAtTime.find(id);
+      TORCH_INTERNAL_ASSERT(
+          it != futuresAtTime.end(),
+          "Error: could not find future in futureTimeouts map, race condition.");
+      futuresAtTime.erase(it);
+      if (futuresAtTime.empty()) {
+        // remove the key from futureTimeouts_
+        futureTimeouts_.erase(endTime);
+      }
+    }
+    futureCV_.notify_all();
+    --clientActiveCalls_;
+    if (message.type() == MessageType::EXCEPTION) {
+      fm->setError(
+          std::string(message.payload().begin(), message.payload().end()));
+    } else {
+      fm->markCompleted(std::move(message));
+    }
+  } else {
+    // TODO: pass the error back to the caller instead of crashing here.
+    TORCH_INTERNAL_ASSERT(false, "unrecognized message type ", message.type());
+  }
+  return 1;
+}
+
+void ProcessGroupAgent::enqueueRecv(RecvWork work) {
+  threadPool_.run(std::bind(
+      [&](RecvWork& work) {
+        try {
+          auto shouldIncr = handleRecv(work);
```

Comment:
nit: (fix can come in the next PR is necessary) Looks like `shouldIncr` is only used here. Does the following work?

```c++
if (handleRecv(work)) {
    recvCounts_.increment(work.from_.id_);
}
```

Reply:
Yes, makes sense. It only needs to be a boolean.